keps/0033-structured-parameters.md

```diff
@@ -0,0 +1,512 @@
+
+`v1beta1`:
+```go
+type InstanceSpec struct {
+    ....
+	Parameters map[string]string `json:"parameters,omitempty"`
+    ....
+}
+```
+
+`v1beta2`:
+```go
+type InstanceSpec struct {
+    ....
+    Parameters apiextension.JSON `json:"parameters"`
+    ....
+}
+```
+
+The existing instance parameter values are currently stored as serialized strings. They will have to be unwrapped and stored as `apiextension.JSON`. The Instance should have accessor or helper functions to convert the `apiextension.JSON` values to their actual types using the JSON-schema provided by the corresponding `OperatorVersion`. 
+
+Here is an example: give a `v1beta1` instance parameters like:
+
+```yaml
+spec:
+  parameters:
+    NODE_COUNT: "3"
+    BACKUP_ENABLED: "false"
+    ARRAY_PARAM: '[foo,bar,bazz]'
+    MAP_PARAM: '{foo:"bar"}'
+```
+
+they will be converted into `v1beta2` and will look like:
+
+```yaml
+spec:
+  parameters:
+    NODE_COUNT: 3
+    BACKUP_ENABLED: false
+    ARRAY_PARAM:
+    - foo
+    - bar
+    - bazz
+    MAP_PARAM: 
+      foo: "bar"
+```
+
+// TODO: it seems that the conversion webhook doesn't have access to the cluster resources (at least by default). How do we unwrap the values (e.g. `MAP_PARAM`) without access to the `OperatorVersion` JSON-schema?
+
+### Client-side
+
+#### Instance updates from the command line (`-p` option)
+
+When updating an operator instance, a user can specify parameters override either in the command line directly (`-p` option) or using an extra file (`-P` option) e.g.: `kudo update --instance my-instance -p KEY=newValue`. This is easy with a flat list of parameters, but becomes more complex with a nested structure. 
+
+The key is now be a [JSON pointer](https://tools.ietf.org/html/rfc6901) expression, the value can either be a simple value, or a more complex json/yaml structure, depending on the specified path. On an update, the existing parameter structure is modified based on the given parameter overrides and only after all updates are applied it is validated against the JSON-schema of the operator.
+
+Examples:
+- set the top-level `clusterName` parameter: 
+  `kudo update ... -p clusterName="NewClusterName"`
+- set the `datacenterName` field of the first item in the `topology` array
+  `kudo update ... -p topology/0/datacenterName="NewDC1"`
+- replace the first item in the `labels` array of the first `topology` array element
+  `kudo update ... -p topology/0/labels/0={ key: "Usage", value: "newValue" }`
+- replace the full labels array of the first topology element
+`kudo update ... -p topology/0/labels=[ { key: "DCLabel", value: "dc1" }, { key: "Usage", value: "additionalItem" } ]`
+
+The `kudo update` command will not allow adding or removing a single entry from an array field, as this would make the command not idempotent anymore.
```

Comment:
so how would you add new entry? Only through the file and by specifying the whole array?

Reply:
Maybe `kudo patch --instance my-instance <command-to-add-array-entry>`? I'm not sure how the exact syntax could or should look like, maybe we can look at the kubectl patching...

For now the only way is to replace the full array, yes.

Reply:
yeah I think that's fine as long as we document this

Reply:
I've added some more details about this.